File: src/main/golang/leetcode/editor/cn/[133]Clone Graph.go
package lc

type Node struct {
	Val       int
	Neighbors []*Node
}

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */

func cloneGraph(node *Node) *Node {
	visited := make(map[*Node]bool)
	mapping := make(map[*Node]*Node)
	var traversal func(node *Node) *Node
	traversal = func(node *Node) *Node {
		if _, ok := visited[node]; ok {
			return mapping[node]
		}
		visited[node] = true
		copied := &Node{Val: node.Val}
		mapping[node] = copied
		for _, neighbor := range node.Neighbors {
			copied.Neighbors = append(copied.Neighbors, traversal(neighbor))
		}
		return copied
	}
	if node == nil {
		return nil
	}
	return traversal(node)
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[133]Clone Graph.py
class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


# leetcode submit region begin(Prohibit modification and deletion)
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""


class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        visited = dict()

        def dfs(node):
            if not node:
                return None
            if node in visited:
                return visited[node]
            clone = Node(node.val, [])
            visited[node] = clone
            for n in node.neighbors:
                clone.neighbors.append(dfs(n))
            return clone
        return dfs(node)
# leetcode submit region end(Prohibit modification and deletion)
